Assets/Scripts/UI/Tooltip.cs

```diff
@@ -0,0 +1,224 @@
+﻿using UnityEngine;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace UI
+{
+    public class Tooltip : MonoBehaviour
+    {
+        [Tooltip("Highlight color when hover over object.")]
+        [SerializeField] private Color highlightColor = new Color(0.38f, 0.97f, 0.44f);
+
+        [Tooltip("Fade speed of the color change (slow -> quick)")]
+        [SerializeField] private float highlightSpeed = 4f;
+
+        [Tooltip("Show a text over the interacted object.")]
+        [SerializeField] private bool hideTooltip;
+
+        [Tooltip("Show a predefined UI Panel over the interacted object (intend to use for items in inventory)")]
+        [SerializeField] private GameObject tooltipUIPanel;
+
+        [Tooltip("Show the tooltip fixed the object instead of following the mouse.")]
+        [SerializeField] private bool fixedToTheObject = true;
+
+        [Tooltip("Position of the tooltip showed over the interacted object.")]
+        [SerializeField] private Vector2 tooltipPosition = new Vector2(-50f, -80f);
+
+        [Tooltip("Text to show over the interacted object.")]
+        [SerializeField] private string tooltipText;
+
+        [Tooltip("Color of the text showed over the interacted object.")]
+        [SerializeField] private Color tooltipColor = new Color(0.9f, 0.9f, 0.9f);
+
+        [Tooltip("Size of the text showed over the interacted object.")]
+        [SerializeField] private int tooltipSize = 20;
+
+        [Tooltip("Resize the text, relative to the distance between the object and the camera.")]
+        [SerializeField] private bool textResized;
+
+        [Tooltip("Font of the text showed over the interacted object.")]
+        [SerializeField] private Font tooltipFont;
+
+        [SerializeField] private enum TooltipAlignment { Center, Left, Right }
+
+        [Tooltip("Alignment of the text showed over the interacted object.")]
+        [SerializeField] private TooltipAlignment tooltipAlignment;
+
+        [Tooltip("Color of the text shadow showed over the interacted object.")]
+        [SerializeField] private Color tooltipShadowColor = new Color(0.1f, 0.1f, 0.1f);
+
+        [Tooltip("Position of the text shadow showed over the interacted object.")]
+        [SerializeField] private Vector2 tooltipShadowPosition = new Vector2(-2f, -2f);
+
+        [Tooltip("Texture for the icon beside tooltip")]
+        [SerializeField] private Texture image;
+
+        // Reference Components
+        private Renderer _renderer;
+        private Material[] allMaterials;
+        private Color[] baseColor;
+        private float t = 0f;
+        private bool over = false;
+        private string currentText = "";
+        private GUIStyle tooltipStyle = new GUIStyle();
+        private GUIStyle tooltipStyleShadow = new GUIStyle();
+        private Vector3 positionToScreen;
+        private float cameraDistance;
+
```

Comment:
I don't really like how the vars are named, and that many of them are initialized at declaration, but I guess I'll shut up

Reply:
I just don't get the idea why won't we define something with default value at declaration (that can be tweaked ofc). It will make our job easier when creating stuffs, testing and shit. Else every time we import the script to an object, we have to set those values by hand again. Imagine you have to do that for 100 different objects.

Reply:
FIrst: I have to agree with Samuel. It is very confusing to have SerializedFields with defined values. When debugging it can easily throw you off.
Second: Prefabs are exactly for this. In there you can define default values and when added to another object you can change/tweak them as you want